include/roboteam_utils/LineSegment.h

```diff
@@ -207,6 +207,19 @@ class LineSegment {
          */
         [[nodiscard]] std::optional<Vector2> nonSimpleIntersects(const LineSegment &line) const;
 
+        /**
+         * @brief Computes the shadow caused by an obstacle (LineSegment) and light source on this LineSegment.
+         * The shadow is a part of this LineSegment. And something is considered to be in the shadow if the line between the source and the position on the line intersects with
+         * the obstacle LineSegment (so if the source is on the obstacle LineSegment then everything is in the shadow).
+         *
+         * @param source The place where the 'light source' is located, i.d. the position from which the projected shadow is computed.
+         * @param obstacle The obstacle LineSegment which blocks the 'lights' and therefore might cause a shadow on this LineSegment.
+         * @param negligible_shadow_length If the length of the shadow is smaller than this value then there is no shadow (if not set then the negligible shadow length is 1e-6).
+         * @return No Linesegment if there is no shadow or if the shadow is smaller than the negligible length. Otherwise return the Linesegment that represents the shadow on this
+         * LineSegment.
+         */
+        [[nodiscard]] std::optional<LineSegment> shadow(const Vector2 &source, const LineSegment &obstacle, float negligible_shadow_length = 1e-6) const;
+
```

Comment:
I really don't see why this is part of this class. This seems like a function that should just exist in some namespace as it has nothing to do with the class linesegment inherently: it is more of a calculation you do on a pair of line segments given a point, and is much more complex in calculation nature than many of the other functions. There is no inherent 'property' or relative property of the linesegment being used here.

Reply:
Sort of agree with you. My vision about Roboteam_utils is that all geometric computations are moved inside it, so no geometric computations in both ControlUtils and in Roboteam_utils. If computations are related to multiple classes then a separate class could be created for it or it could be defined in multiple classes with a reference to the corresponding class where it is defined.